src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -13690,6 +13690,9 @@ void Unit::UpdateCharmAI()
             delete i_AI;
             i_AI = i_disabledAI;
             i_disabledAI = NULL;
+
+            if (GetTypeId() == TYPEID_UNIT)
```

Comment:
That is the point. RemoveCharmBy's call is handled by the charmed AI, while this call will be handled by the creature's _original_ AI.